GCEWindowsAgent/oslogin.go

```diff
@@ -0,0 +1,339 @@
+	}
+	return nil
+}
+
+func updatePAMConfig(enable, twofactor bool) error {
+	authOSLogin := "auth       [success=done perm_denied=die default=ignore] pam_oslogin_login.so"
+	authGroup := "auth       [default=ignore] pam_group.so"
+	accountOSLogin := "account    [success=ok ignore=ignore default=die] pam_oslogin_login.so"
+	accountOSLoginAdmin := "account    [success=ok default=ignore] pam_oslogin_admin.so"
+	sessionHomeDir := "session    [success=ok default=ignore] pam_mkhomedir.so"
+
+	if runtime.GOOS == "freebsd" {
+		authOSLogin = "auth       optional pam_oslogin_login.so"
+		authGroup = "auth       optional pam_group.so"
+		accountOSLogin = "account    requisite pam_oslogin_login.so"
+		accountOSLoginAdmin = "account    optional pam_oslogin_admin.so"
+		sessionHomeDir = "session    optional pam_mkhomedir.so"
+	}
+
+	pamsshd, err := ioutil.ReadFile("/etc/pam.d/sshd")
+	if err != nil {
+		return err
+	}
+	filtered := filterGoogleLines(string(pamsshd))
+	if enable {
+		topOfFile := []string{googleBlockStart}
+		if twofactor {
+			topOfFile = append(topOfFile, authOSLogin)
+		}
+		topOfFile = append(topOfFile, []string{authGroup, googleBlockEnd}...)
+		bottomOfFile := []string{googleBlockStart, accountOSLogin, accountOSLoginAdmin, sessionHomeDir, googleBlockEnd}
+		filtered = append(topOfFile, filtered...)
+		filtered = append(filtered, bottomOfFile...)
+	}
+	proposed := strings.Join(filtered, "\n")
+	if proposed != string(pamsshd) {
+		file, err := os.OpenFile("/etc/pam.d/sshd", os.O_WRONLY|os.O_TRUNC, 0777)
+		if err != nil {
+			return err
+		}
+		defer file.Close()
+		file.WriteString(proposed)
+	}
+
+	accountSu := "account    [success=bad ignore=ignore] pam_oslogin_login.so"
+
+	pamsu, err := ioutil.ReadFile("/etc/pam.d/su")
+	if err != nil {
+		return err
+	}
+	filtered = filterGoogleLines(string(pamsu))
+	if enable {
+		filtered = append([]string{googleComment, accountSu}, filtered...)
+	}
+	proposed = strings.Join(filtered, "\n")
+	if proposed != string(pamsu) {
+		file2, err := os.OpenFile("/etc/pam.d/su", os.O_WRONLY|os.O_TRUNC, 0777)
+		if err != nil {
+			return err
+		}
+		defer file2.Close()
+		file2.WriteString(proposed)
+	}
+
+	return nil
+}
+
+func createOSLoginDirs() error {
+	restorecon, err := exec.LookPath("restorecon")
+	if err != nil {
+		restorecon = ""
+	}
+
+	for _, dir := range []string{"/var/google-sudoers.d", "/var/google-users.d"} {
+		err := os.Mkdir(dir, 0750)
+		if err != nil && !os.IsExist(err) {
+			return err
+		}
+		if restorecon != "" {
+			exec.Command(restorecon, dir).Run()
+		}
+	}
+	return nil
+}
+
+func createOSLoginSudoersFile() error {
+	osloginSudoers := "/etc/sudoers.d/google-oslogin"
+	if runtime.GOOS == "freebsd" {
+		osloginSudoers = "/usr/local/etc/sudoers.d/google-oslogin"
+	}
+	sudoFile, err := os.OpenFile(osloginSudoers, os.O_WRONLY|os.O_CREATE|os.O_EXCL, 0440)
+	if err != nil {
+		if os.IsExist(err) {
+			return nil
+		}
+		return err
+	}
+	defer sudoFile.Close()
```

Comment:
check the error, for this its simple to just return sudoFile.Close() as you have no other returns